[Traceable FSDP2][Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager

yf225 opened the pull request:
Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138187
* #138241
* __->__ #138113



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @xmfan

yf225 committed:
[Compiled Autograd] Check stance to decide whether to fallback to eager

[ghstack-poisoned]

yf225 committed:
Update on "[Compiled Autograd] Check stance to decide whether to fallback to eager"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 committed:
Update on "[Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager"

Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge -f "stuck in CI, also low-risk to merge since Dynamo stance is a new API"

yf225 commented:
@pytorchbot merge -f "stuck waiting for inductor_torchbench_smoketest_perf job which doesn't test this path, also low-risk to merge since Dynamo stance is a new API"

clee2000 commented:
@pytorchbot revert -m "sorry need to revert this in order to revert #137953, please rebase and remerge" -c werid

clee2000 commented:
@pytorchbot revert -m "sorry need to revert this in order to revert https://github.com/pytorch/pytorch/pull/137953, please rebase and remerge" -c weird

yf225 commented:
@pytorchbot rebase

yf225 committed:
Update on "[Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager"

Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge

yf225 committed:
Update on "[Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager"

Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge

yf225 committed:
Update on "[Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager"

Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 committed:
Update on "[Compiled Autograd] Check Dynamo stance to decide whether to fallback to eager"

Dynamo stance is recently added in https://github.com/pytorch/pytorch/pull/137504. When Dynamo stance is "force_eager" (user explicitly wants to fall back to eager), we would like Compiled Autograd to fall back to eager as well. This will allow the Traceable FSDP2 use case to work since "eager forward + compiled autograd backward" is not supported for Traceable FSDP2.

In general, if user wants to do "eager forward + compiled autograd backward", they should explicitly run the forward in eager instead of applying compile and then set stance to "force_eager".




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec xmfan

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge -f "stuck in inductor_torchbench_smoketest_perf CI job, also low-risk to merge since Dynamo stance is a new API"

yf225 commented:
@pytorchbot merge -f "stuck in inductor_torchbench_smoketest_perf CI job, also low-risk to merge since Dynamo stance is a new API"
